File: LeetCode/LeetCode-CPP/2020JuneLeetCodingChallenge/Cheapest Flights Within K Stops.cpp
class Solution {
public:
    int findCheapestPrice(int n, vector<vector<int>>& flights, int src, int dst, int K) {
        // bellman ford
        vector<int> c(n,INT_MAX);
        c[src] = 0;
        // 停顿K次，则需要K+1个边
        for(int i = 0; i <= K; i++) {
            vector<int> temp(c); // 拷贝一份数据用来计算
            for(auto e : flights) {
                // 到达下一个点的距离 = min(目前到达下一个点的距离，原来到达中转点+ 中转点到下一个点 的距离 );
                int k = c[e[0]] == INT_MAX ? INT_MAX : c[e[0]] + e[2];
                temp[e[1]] = min(temp[e[1]], k); // 遍历所有的边，计算出到达下一个点的最小费用
            }
                
            c = temp; // 更新距离
        }
        return c[dst] == INT_MAX ? -1 : c[dst];
    }
};

File: LeetCode/LeetCode-Java/2020JuneLeetCodingChallenge/Cheapest Flights Within K Stops.java
class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int K) {
        // bellman ford
        int[] prices = new int[n];
        Arrays.fill(prices, Integer.MAX_VALUE);
        prices[src] = 0; // 将src设置为0
        for(int i = 0; i <= K; i++) {
            // 需要单独复制一份用来查看原来的记录，来限制一次只走一步，如果修改原来的数组，则遍历完所有的边相当于走了很多步.eg:0->1 100, 1->2 100,0->2 500, 求0->2 且K=0.
            int[] temp = Arrays.copyOf(prices, n);
            for(int[] e : flights) {
                if(prices[e[0]] != Integer.MAX_VALUE) { // 如果==MAX则说明之前无法到达该中转点，所以无需比较
                    temp[e[1]] = Math.min(temp[e[1]], prices[e[0]] + e[2]);
                }
            }
            prices = temp;
        }
        return prices[dst] == Integer.MAX_VALUE ? -1 : prices[dst];
    }
}